### ISSUE TITLE
everywhere() sometimes fails to reach everywhere when using dispatcher

### ISSUE THREAD

**Issue by gangstR** [shikokuchuo/mirai#132]
I'd previously tried to substantiate the case that there is an issue here to no avail. You can read more in my [last comment on my previous issue that was downgraded to the discussion forum](https://github.com/shikokuchuo/mirai/discussions/131#discussioncomment-10153919). It has more details and commentary along with a slightly different example.

It is clear to me that `everywhere()` is not reliable for reproducible work when placed under even moderate load. Using the same example code supplied in the documentation, you can readily break the behavior by exerting a bit of pressure. In the example code, the `parallel` package that is included in the base R distribution is loaded and attached to the search path via the call `everywhere(library(parallel))`. Subsequently, a `mirai("package:parallel" %in% search())` call is made and that mirai is resolved with a call to `call_mirai()`. That may work well on the single daemon that was started; however, try expanding that example to execute many times across a larger set of daemons and witness intermittent failures over iterations of that process:

```{r}
library(mirai)
cnt <- 1L
while (TRUE) {
  daemons(10)
  everywhere(library(parallel))
  m.vec <- unlist(lapply(1:1000, FUN = function(i) {
    m <- mirai("package:parallel" %in% search())
    resolved.mirai <- call_mirai_(m)
    return(resolved.mirai$data)
  }))
  not.really.resolved.mirai.cnt <- sum(!m.vec)
  if (not.really.resolved.mirai.cnt > 0) {
    message("I'm not reproducible on run #", cnt, " because "
            , not.really.resolved.mirai.cnt
            , " time(s) out of 1000 I was not really ready :'-(")
  }
  daemons(0)
  cnt <- cnt + 1L
} 
```

Be patient as it may take a few tries, but eventually you'll see failures. Here are some of my own:

```
I'm not reproducible on run #8 because 126 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #26 because 316 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #50 because 46 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #55 because 76 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #57 because 92 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #60 because 567 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #65 because 101 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #80 because 380 time(s) out of 1000 I was not really ready :'-(
I'm not reproducible on run #85 because 571 time(s) out of 1000 I was not really ready :'-(
```

To me, what I'm doing above is not an unreasonable extension to the example that was supplied to test the behavior _under pressure_ which is how all production work should be evaluated. Per my [last comment in the previous issue I tried to open](https://github.com/shikokuchuo/mirai/discussions/131#discussioncomment-10153919), this is either a bug in the code or a bug in the documentation because the function does not do what it claims to do. You might think this is an artifact of the pressure exerted here of pushing a 1000 mirai onto 10 daemons, but that's not the case. If you reduce the `lapply()` to only `10` iterations to match the daemon count, it will still fail intermittently -- you'll just need to be more patient since it becomes less likely to happen.

As it is, users should be prepared for unexpected failures in scripts for any non-trivial setup work at scale.

Thank you

**Comment by shikokuchuo** [comment]
This has nothing to do with load - if I understand correctly you are simply launching lots of mirai at the same time to see if the `everywhere()` call has registered on all daemons. If it is indeed the case that it has not then it's simply a bug in `everywhere()`.

I cannot get this to reproduce, so (i) have you managed to get it to occur on other machines, and (ii) what are the package versions you are using, the R version, OS, also the value of `nanonext::nng_version()`?

Also, if you replace the daemons call with `daemons(10, dispatcher = FALSE)`, does this also run into the issues or not? That will help determine what I'm looking at.

Thanks.

**Comment by gangstR** [comment]
Thank you for the response. Yes, what I've been saying is that I can generate clear evidence of a problem that manifests within `everywhere()`. The real defect could be within the dispatcher, but, if so, it thus far is only manifesting in `everywhere()` and not from within any other function in your package -- at least that I can detect so far. 

**Have I reproduced the issue?**

Yes, two different machines, two Windows versions, two R versions, and two versions of your respective packages built from source (details follow).

HP Elite Dragonfly
Intel Core i7-1185G7 @ 3.00GHz
32GB RAM
Windows 10

Version Info
```
>nanonext::nng_version()
[1] "1.8.0"          "Mbed TLS 3.5.2"
>packageVersion("mirai")
[1] ‘1.1.1.9002’
>packageVersion("nanonext")
[1] ‘1.1.1.9005’

>sessionInfo()
R version 4.2.3 (2023-03-15 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] nanonext_1.1.1.9005 mirai_1.1.1.9002   

loaded via a namespace (and not attached):
 [1] compiler_4.2.3    fastmap_1.2.0     cli_3.6.1         htmltools_0.5.8.1 tools_4.2.3      
 [6] parallel_4.2.3    rstudioapi_0.15.0 yaml_2.3.8        rmarkdown_2.18    knitr_1.41       
[11] xfun_0.35         digest_0.6.30     rlang_1.1.4.9000  evaluate_0.18   
```

Dell XPS 9570
Intel Core i7-8750-H @ 2.2GHz
16GB RAM
Windows 11

Version info
```
>nanonext::nng_version()
[1] "1.8.0"          "Mbed TLS 3.5.2"
>packageVersion("mirai")
[1] ‘1.1.1.9008’
>packageVersion("nanonext")
[1] ‘1.1.1.9010’

> sessionInfo()
R version 4.4.1 (2024-06-14 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] nanonext_1.1.1.9010 mirai_1.1.1.9008   

loaded via a namespace (and not attached):
[1] compiler_4.4.1   rsconnect_0.8.16 tools_4.4.1  
```

**Evidence**

The problem only reveals itself when a bit of pressure is put on the system. I'm not talking about CPU load; rather basic stress-testing type approaches using many iterations and continuous cycling of the process that will be happening in a production context. For anything going into production for my work, it's going to be tested with an assortment of stress tests and negative tests. 

- If I run your code example for `everywhere()` verbatim, it works 100% of the time. I could not it get it to fail once after running it non-stop overnight. I could try it for longer periods, yes.
- If I run my `while()` variation in the reprex with 10 daemons and 10 iterations in the `lapply()` call, it happens extremely rarely on either machine. I let it run for an hour on machine 1 (HP) and it only had 4 failures -- which is enough evidence of an issue (not a question).
- If I run my `while...` reprex with 10 daemons and a 1000 iterations on machine 1 (HP) machine, it fails at least once 100% of the time within ~90 seconds and frequently at least once in 30 seconds. On machine 2 (Dell), I could elicit similar behavior/timing but it required putting a bit more pressure into the test using more daemons (10 to 15) and more iterations (from 1000 to 1500). Those small changes were sufficient. Admittedly, machine 1 (HP) had far more going on in the background than machine 2 (Dell) which was running nothing other than this test.
- If I use more complex `.expr` arguments in the call to `everywhere()` -- such as multiple provisioning commands that I need to do in my work -- I can also significantly increase the likelihood of the failure to happen. In fact, if I run my own scripts, it will fail about 75% of the time on a single run -- no cycling required.

**Any evidence on another OS?**
- No, not yet. As soon as I have time, I'm going to try to repeat this on Alma in AWS, which is where I normally work.

**What about the dispatcher? Have I tested that setting?**
 - Yes! If I set `dispatcher = FALSE`, I have thus far not been able to generate a failure. Machine 2 (Dell) cycled the test for 18 hours and never generated a single failure with this setting. While I wanted to use the dispatcher with my usage context, I appreciate you asking about this specifically since it may help in isolating the true source of the buggy behavior that is manifesting in `everywhere()`.

Thank you

**Comment by shikokuchuo** [comment]
Thanks for testing on another machine, that's helpful. Also it's fine I don't require you to test on other platforms. If it's reproducible on Windows then I'll know to try there when I next work on that platform.

Given it's contained to `everywhere()` using dispatcher daemons, I have a pretty good idea of what this is. FYI this part of the code was due for an upgrade anyway, but I'll prioritize this now that I know that this can cause issues 'in the wild'.

I may not be able to get onto this immediately, but I'll be sure to check back on this before the next package release.

**Comment by shikokuchuo** [comment]
This issue is now resolved as of mirai 1.1.1.9010.
Note: the behaviour of `everywhere()` has not been changed - that would require a lot more thought.
